http/messages.go

```diff
@@ -0,0 +1,149 @@
+package http
+
+import (
+	"sort"
+
+	"github.com/kkrull/gohttp/msg/clienterror"
+	"github.com/kkrull/gohttp/msg/servererror"
+)
+
+func NewGetMessage(path string) RequestMessage {
+	return &requestMessage{
+		method: "GET",
+		target: path,
+		path:   path,
+	}
+}
+
+func NewHeadMessage(path string) RequestMessage {
+	return &requestMessage{
+		method: "HEAD",
+		target: path,
+		path:   path,
+	}
+}
+
+func NewOptionsMessage(target string) RequestMessage {
+	return &requestMessage{
+		method: "OPTIONS",
+		target: target,
+		path:   target,
+	}
+}
+
+func NewPutMessage(path string) RequestMessage {
+	return &requestMessage{
+		method: "PUT",
+		target: path,
+		path:   path,
+	}
+}
+
+func NewTraceMessage(path string) RequestMessage {
+	return &requestMessage{
+		method: "TRACE",
+		target: path,
+		path:   path,
+	}
+}
+
+func NewRequestMessage(method, path string) RequestMessage {
+	return &requestMessage{
+		method: method,
+		target: path,
+		path:   path,
+	}
+}
+
+type requestMessage struct {
+	method          string
+	path            string
+	target          string
+	queryParameters []QueryParameter
+}
+
+func (message *requestMessage) Method() string {
+	return message.method
+}
+
+func (message *requestMessage) Path() string {
+	return message.path
+}
+
+func (message *requestMessage) AddQueryFlag(name string) {
+	message.queryParameters = append(message.queryParameters, QueryParameter{Name: name})
+}
+
+func (message *requestMessage) AddQueryParameter(name, value string) {
+	message.queryParameters = append(message.queryParameters, QueryParameter{Name: name, Value: value})
+}
+
+func (message *requestMessage) QueryParameters() []QueryParameter {
+	return message.queryParameters
+}
+
+func (message *requestMessage) Target() string {
+	return message.target
+}
+
+func (message *requestMessage) NotImplemented() Response {
+	return &servererror.NotImplemented{Method: message.method}
+}
+
+func (message *requestMessage) MakeResourceRequest(resource Resource) Request {
+	if message.method == "OPTIONS" {
+		return &optionsRequest{
+			SupportedMethods: message.supportedMethods(resource),
+		}
+	}
+
+	method := knownMethods[message.method]
+	if method == nil {
+		return message.unknownHttpMethod(resource)
+	}
+
+	request := method.MakeRequest(message, resource)
```

Comment:
That's clever - I hadn't thought about using multiple returns to help clients guard against missing guard clauses.  I followed the model of indexing a map, so it can read

```
request, isSupported := message.MakeRequest(...)
if !isSupported { ...guard clause... }
```